### PR TITLE
Fix how extensions are specified and used in shaders

### PR DESCRIPTION
**Pull request by shehzan10**
The correct syntax of using an extension is:
```
extension GL_EXT_some_ext : enable
// code
#ifdef GL_EXT_some_ext
// code that uses the extension
endif // GL_EXT_some_ext
```

Ref #4711 and #2860 
Note: contain trailing whitespace fixes.

-------
### More Failing Tests
There are a few more tests that fail because of the extensions. One example is [Context/get fragment depth extension](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Specs/Renderer/ContextSpec.js#L191-L232). Similar to the ones listed in #4711.

The problem with these tests is that they run 2 shaders. The first is valid and does not require any extensions and sets the color of the pixel (lets say to `x`).
The second shader uses the extension, but also creates a valid shader in case of the extension not being available. However, the [expectation](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Specs/Renderer/ContextSpec.js#L225) is that if the extension is unavailable, the second shader won't run and the first's shader's value should be expected. However, because the second shader is also being generated correctly, its going to run fine.

So the way to fix these types of tests is to do one of 2 things:
1. Always expect the pixel color from the second shader.
2. Create the second shader with the extension and it will fail without the extension being unavailable and match the color of the 1st shader.

If my explanation is correct, I can go ahead and fix these tests too.

### CODE REVIEW

File: Source/Shaders/AdjustTranslucentFS.glsl [hunk 1]
```diff
@@ -11,14 +11,14 @@ void main()
 {
     if (texture2D(u_depthTexture, v_textureCoordinates).r < 1.0)
     {
-#ifdef MRT
+#ifdef GL_EXT_draw_buffers
```
Comment (shehzan10) [inline Source/Shaders/AdjustTranslucentFS.glsl]: A safer version of this is `#if defined(MRT) && defined(GL_EXT_draw_buffers)`. Should I use that?

The only way `GL_EXT_draw_buffers` should be enabled is if `MRT` is defined above. So this *should* be enough.
But if implementation was incorrect and the extensions are always enabled, then is would no longer be equivalent.

File: Source/Shaders/Materials/PolylineArrowMaterial.glsl [hunk 2]
```diff
@@ -13,23 +13,28 @@ float getPointOnLine(vec2 p0, vec2 p1, float x)
...
     czm_material material = czm_getDefaultMaterial(materialInput);
-    
+
     vec2 st = materialInput.st;
-    
+
+#ifdef GL_OES_standard_derivatives
     float base = 1.0 - abs(fwidth(st.s)) * 10.0;
+#else
+    float base = 1.0;  // FIXME
```
Comment (shehzan10) [inline Source/Shaders/Materials/PolylineArrowMaterial.glsl]: Can someone recommend a proper fallback computation for `fwidth`?

**Comment by pjcozzi** [comment]
> So the way to fix these types of tests is to do one of 2 things:
>
> Always expect the pixel color from the second shader.
Create the second shader with the extension and it will fail without the extension being unavailable and match the color of the 1st shader.
>
>If my explanation is correct, I can go ahead and fix these tests too.

For cohesion, it would be fine to check for the extension at the top of the test and just `return` if it is not supported.  Several tests do this elsewhere.

File: Source/Shaders/AdjustTranslucentFS.glsl (hunk 1 above)
Comment (pjcozzi) [inline Source/Shaders/AdjustTranslucentFS.glsl]: Couldn't this stay as `#ifdef MRT`?  If the extension is not supported, the JavaScript code concatenating the shader should not `#defined MRT`.

**Comment by pjcozzi** [comment]
> The correct syntax of using an extension is:
```
extension GL_EXT_some_ext : enable
// code
#ifdef GL_EXT_some_ext
// code that uses the extension
endif // GL_EXT_some_ext
```

Are you sure?  Where did you see this?

Here's what the GLSL spec says:
> Macro expansion is not done on lines containing #extension and #version directives.
For each extension there is an associated macro. The macro is always defined in an implementation that
supports the extension. This allows the following construct to be used:
#ifdef OES_extension_name
#extension OES_extension_name : enable
// code that requires the extension
#else
// alternative code
#endif

https://www.khronos.org/files/opengles_shading_language.pdf

It's possible that there are bugs in WebGL implementations or the GLSL version is out of sync.

File: Source/Shaders/Materials/PolylineArrowMaterial.glsl (hunk 2 above)
Comment (pjcozzi) [inline Source/Shaders/Materials/PolylineArrowMaterial.glsl]: It requires screen-space partial derivatives, which we really can't emulate.

Without it, the arrow head will likely be sized in meters, not pixels, which could be an OK fallback.  Try it and see.

**Comment by shehzan10** [comment]
@pjcozzi I fixed the extensions and there was actually just one fix to be made (as discussed offline).

I have also disabled the frame dept test if the extension is unavailable

File: Source/Shaders/Materials/PolylineArrowMaterial.glsl (hunk 2 above)
Comment (pjcozzi) [inline Source/Shaders/Materials/PolylineArrowMaterial.glsl]: We generally don't merge code with `// FIXME` or `TODO` comments into master (except for performance ideas).

What is the plan here?  Did you test this fallback and is it OK with you?

**Comment by pjcozzi** [comment]
At this point, this no longer fixes #4711 or #2860, right?

It's still a fine change, I'm just trying to understand its impact.

**Comment by shehzan10** [comment]
> At this point, this no longer fixes #4711 or #2860, right?
> 
> It's still a fine change, I'm just trying to understand its impact.

Actually it does. This fixes the error
```
Scene/GroundPrimitive renders with distance display condition per instance attribute
DeveloperError: must call update before calling getGeometryInstanceAttributes
Error
    at new DeveloperError (http://localhost:8080/Source/Core/DeveloperError.js:44:19)
    at GroundPrimitive.getGeometryInstanceAttributes (http://localhost:8080/Source/Scene/GroundPrimitive.js:1195:19)
    at http://localhost:8080/Specs/Scene/GroundPrimitiveSpec.js:543:40
    at Object.<anonymous> (http://localhost:8080/Specs/spec-main.js:139:30)
    at attemptAsync (http://localhost:8080/ThirdParty/jasmine-2.2.0/jasmine.js:1793:24)
    at QueueRunner.run (http://localhost:8080/ThirdParty/jasmine-2.2.0/jasmine.js:1746:26)
    at QueueRunner.execute (http://localhost:8080/ThirdParty/jasmine-2.2.0/jasmine.js:1732:10)
    at Spec.queueRunnerFactory (http://localhost:8080/ThirdParty/jasmine-2.2.0/jasmine.js:620:35)
    at Spec.execute (http://localhost:8080/ThirdParty/jasmine-2.2.0/jasmine.js:357:10)
    at Object.fn (http://localhost:8080/ThirdParty/jasmine-2.2.0/jasmine.js:2095:43)
Scene/Material draws PolylineArrow built-in material
RuntimeError: Fragment shader failed to compile.  Compile log: WARNING: 0:? : '
' : extension directive should occur before any non-preprocessor tokens 
WARNING: 0:? : 'GL_OES_standard_derivatives' : extension is not supported
ERROR: 0:18: 'fwidth' : no matching overloaded function found 
```

Which is listed in both those issues.

**Comment by pjcozzi** [comment]
OK, we'll still leave those issues open.  Please bump this issue once you have a plan for `fwidth` - it will only impact a very small percentage of users so it doesn't need to be perfect.

**Comment by shehzan10** [comment]
`fwidth` is being used to compute the base, which is the ratio of the length of the line to the arrow head. (0.999999 will mean a very small arrow head, and 0 would mean the entire line will look like a arrow head).

In my experimentation, I have found `0.99` to be a good ratio. This change has been pushed.

**Comment by pjcozzi** [comment]
Thanks for all the detective work, @shehzan10!

**Pull request merged**